File: go_struct/hero_quest/hero_quest.go
package hero_quest

import "fmt"

type QuestData struct {
	PlayerName          string
	PlayerHealth        int
	PlayerStrength      int
	PlayerMagic         int
	PlayerCraftingSkill int
	ItemName            string
	ItemKind            string
	ItemPower           int
}

func PlayerToString(playerName string, playerHealth int, playerStrength int, playerMagic int, playerCraftingSkill int) string {
	return fmt.Sprintf(
		"%s's Attributes:\nHealth: %v\nStrength: %v\nMagic: %v\nCrafting Skill: %v\n",
		playerName,
		playerHealth,
		playerStrength,
		playerMagic,
		playerCraftingSkill,
	)
}

func PlayerFallsDown(questData *QuestData) {
	fmt.Println("Player drops off a cliff.")

	if questData.PlayerStrength < 5 {
		questData.PlayerHealth -= 10
		fmt.Println("Player's strength is too small. Health decreases by 10.")
	}
}

func ItemToString(itemName string, itemKind string, itemPower int) string {
	return fmt.Sprintf("Item: %s\nKind: %s\nPower: %v\n", itemName, itemKind, itemPower)
}

func ItemReduceByUsage(questData *QuestData) {
	fmt.Printf("Using the item with kind '%s' and power %v\n", questData.ItemKind, questData.ItemPower)

	questData.ItemPower = questData.ItemPower / 2
	if questData.ItemPower == 0 {
		questData.ItemKind = "Junk"
	}
}

func ItemApplyEffectToPlayer(questData *QuestData) {
	fmt.Printf("Applying the effect of %s (%s):\n", questData.ItemName, questData.ItemKind)
	switch questData.ItemKind {
	case "Health":
		questData.PlayerHealth += questData.ItemPower
	case "Strength":
		questData.PlayerStrength += questData.ItemPower
	case "Magic":
		questData.PlayerMagic += questData.ItemPower
	}
}

func ItemRepair(questData *QuestData) {
	fmt.Println("Using the repair skill to fix the item:")

	repairAmount := -5 + ((questData.PlayerCraftingSkill * 2) + 1)

	questData.ItemPower += repairAmount

	fmt.Printf("Repaired the item by %v points. Item's Durability: %v\n", repairAmount, questData.ItemPower)
}

File: PHP/src/HeroQuest.php
<?php

namespace CodingDojo;

class HeroQuest {

    public static function playerToString($playerName, $playerHealth, $playerStrength, $playerMagic, $playerCraftingSkill) {
        return sprintf("%s's Attributes:\nHealth: %d\nStrength: %d\nMagic: %d\nCrafting Skill: %d\n",
            $playerName, $playerHealth, $playerStrength, $playerMagic, $playerCraftingSkill);
    }

    public static function playerFallsDown(&$playerHealth, $playerStrength) {
        echo "Player drops off a cliff.\n";

        if ($playerStrength < 5) {
            $playerHealth -= 10;
            echo "Player's strength is too small. Health decreases by 10.\n";
        }
    }

    public static function itemToString($itemName, $itemKind, $itemPower) {
        return sprintf("Item: %s\nKind: %s\nPower: %d\n", $itemName, $itemKind, $itemPower);
    }

    public static function itemReduceByUsage(&$itemKind, &$itemPower) {
        echo sprintf("Using the item with kind '%s' and power %d\n", $itemKind, $itemPower);

        $itemPower = intdiv($itemPower, 2);

        if ($itemPower == 0) {
            $itemKind = "Junk";
        }
    }

    public static function itemApplyEffectToPlayer($itemName, $itemKind, $itemPower, &$playerHealth, &$playerStrength, &$playerMagic) {
        echo sprintf("Applying the effect of %s (%s):\n", $itemName, $itemKind);

        if ($itemKind === "Health") {
            $playerHealth += $itemPower;
        } elseif ($itemKind === "Strength") {
            $playerStrength += $itemPower;
        } elseif ($itemKind === "Magic") {
            $playerMagic += $itemPower;
        } else {
            // ignore unknown item kind
        }
    }

    public static function itemRepair($playerCraftingSkill, &$itemPower) {
        echo "Using the repair skill to fix the item:\n";

        $repairAmount = -5 + (($playerCraftingSkill * 2) + 1);
        $itemPower += $repairAmount;

        echo sprintf("Repaired the item by %d points. Item's Durability: %d\n", $repairAmount, $itemPower);
    }
}
